File: Motocitizen/src/motocitizen/Activity/DetailMessagesFragment.java
package motocitizen.Activity;

import android.net.Uri;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.PopupWindow;
import android.widget.Toast;

import org.json.JSONException;
import org.json.JSONObject;

import motocitizen.accident.Message;
import motocitizen.app.general.popups.MessagesPopup;
import motocitizen.app.general.user.Role;
import motocitizen.content.Content;
import motocitizen.draw.Rows;
import motocitizen.draw.Sort;
import motocitizen.main.R;
import motocitizen.network.requests.AsyncTaskCompleteListener;
import motocitizen.network.requests.SendMessageRequest;

public class DetailMessagesFragment extends AccidentDetailsFragments {

    private OnFragmentInteractionListener mListener;

    private ImageButton newMessageButton;
    private EditText    mcNewMessageText;
    private View        newMessageArea;
    private View        mcDetMessagesTable;

    public static DetailMessagesFragment newInstance(int accID, String userName) {
        DetailMessagesFragment fragment = new DetailMessagesFragment();
        Bundle                 args     = new Bundle();
        args.putInt(ACCIDENT_ID, accID);
        args.putString(USER_NAME, userName);
        fragment.setArguments(args);
        return fragment;
    }

    public DetailMessagesFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, final ViewGroup container, Bundle savedInstanceState) {
        View viewMain = inflater.inflate(R.layout.fragment_detail_messages, container, false);

        newMessageButton = (ImageButton) viewMain.findViewById(R.id.mc_new_message_send);
        newMessageButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String text = mcNewMessageText.getText().toString();
                new SendMessageRequest(new SendMessageCallback(), getActivity(), accidentID, text);
                newMessageButton.setEnabled(false);
            }
        });

        newMessageButton.setEnabled(false);

        mcNewMessageText = (EditText) viewMain.findViewById(R.id.mc_new_message_text);
        mcNewMessageText.addTextChangedListener(mcNewMessageTextListener);

        mcDetMessagesTable = viewMain.findViewById(R.id.mc_det_messages_table);

        newMessageArea = viewMain.findViewById(R.id.mc_new_message_area);

        update();

        return viewMain;
    }

    private void update() {
        ViewGroup messageView = (ViewGroup) mcDetMessagesTable;
        messageView.removeAllViews();
        final motocitizen.accident.Accident accident  = ((AccidentDetailsActivity) getActivity()).getCurrentPoint();
        int                                  lastOwner = 1;
        int                                  nextOwner = 0;

        Integer[] keys = Sort.getSortedMessagesKeys(accident.getMessages());
        for (int i = 0; i < keys.length; i++) {
            if (i < keys.length - 1) {
                nextOwner = accident.getMessages().get(keys[i + 1]).getOwnerId();
            } else nextOwner = 0;
            final Message message = accident.getMessages().get(keys[i]);
            message.setUnread(false);
            View row = Rows.getMessageRow(getActivity(), messageView, message, lastOwner, nextOwner);
            lastOwner = accident.getMessages().get(keys[i]).getOwnerId();
            row.setOnLongClickListener(new View.OnLongClickListener() {

                @Override
                public boolean onLongClick(View v) {
                    PopupWindow popupWindow;
                    popupWindow = (new MessagesPopup(getActivity(), message.getId(), accident.getId())).getPopupWindow();
                    int viewLocation[] = new int[2];
                    v.getLocationOnScreen(viewLocation);
                    popupWindow.showAtLocation(v, Gravity.NO_GRAVITY, viewLocation[0], viewLocation[1]);
                    return true;
                }
            });
            messageView.addView(row);
        }
        setupAccess();
    }

    public void notifyDataSetChanged() {
        update();
        mcNewMessageText.setText("");
    }

    private final TextWatcher mcNewMessageTextListener = new TextWatcher() {

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            String temp = s.toString().replaceAll("\\s", "");
            if (temp.length() == 0) {
                newMessageButton.setEnabled(false);
            } else {
                newMessageButton.setEnabled(true);
            }
        }

        @Override
        public void afterTextChanged(Editable s) {
        }
    };

    // TODO: Rename method, update argument and hook method into UI event
    public void onButtonPressed(Uri uri) {
        if (mListener != null) {
            mListener.onFragmentInteraction(uri);
        }
    }

    private void setupAccess() {
        if (Role.isStandart()) {
            newMessageArea.setVisibility(View.VISIBLE);
        } else {
            newMessageArea.setVisibility(View.INVISIBLE);
        }
    }
    private void message(String text) {
        Toast.makeText(getActivity(), text, Toast.LENGTH_LONG).show();
    }
    private class SendMessageCallback implements AsyncTaskCompleteListener {
        @Override
        public void onTaskComplete(JSONObject result) {
            if (result.has("error")) {
                try {
                    message(result.getString("error"));
                } catch (JSONException e) {
                    message("Неизвестная ошибка" + result.toString());
                    e.printStackTrace();
                }
            } else {
                //new AccidentsRequest(getActivity(), new UpdateAccidentsCallback());
                Content.update(getActivity(), new UpdateAccidentsCallback());
            }
            newMessageButton.setEnabled(true);
        }
    }

    private class UpdateAccidentsCallback implements AsyncTaskCompleteListener {
        @Override
        public void onTaskComplete(JSONObject result) {
            mcNewMessageText.setText("");
            if (!result.has("error")) Content.parseJSON(getActivity(), result);
            ((AccidentDetailsActivity) getActivity()).update();
            update();
        }
    }

}

File: Motocitizen/src/motocitizen/ui/fragments/DetailMessagesFragment.kt
package motocitizen.ui.fragments

import android.app.Fragment
import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageButton
import android.widget.PopupWindow
import android.widget.ScrollView
import motocitizen.content.Content
import motocitizen.content.accident.Accident
import motocitizen.content.message.Message
import motocitizen.datasources.database.StoreMessages
import motocitizen.datasources.network.ApiResponse
import motocitizen.datasources.network.requests.SendMessageRequest
import motocitizen.main.R
import motocitizen.ui.activity.AccidentDetailsActivity
import motocitizen.ui.activity.AccidentDetailsActivity.Companion.ACCIDENT_ID_KEY
import motocitizen.ui.menus.MessageContextMenu
import motocitizen.ui.rows.message.MessageRowFactory
import motocitizen.user.User
import motocitizen.utils.hide
import motocitizen.utils.show
import motocitizen.utils.showToast
import org.json.JSONException

class DetailMessagesFragment() : Fragment() {
    private lateinit var rootView: View
    private lateinit var scrollView: ScrollView
    private lateinit var messagesView: ViewGroup
    private lateinit var formView: View
    private lateinit var sendButton: ImageButton
    private lateinit var inputField: EditText
    private lateinit var accident: Accident

    constructor(accident: Accident) : this() {
        this.accident = accident
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup, savedInstanceState: Bundle?): View? {
        rootView = inflater.inflate(R.layout.fragment_detail_messages, container, false)
        bindViews()
        setupNewMessageForm()

        update()//todo exterminatus

        return rootView
    }

    private fun bindViews() {
        scrollView = rootView.findViewById(R.id.activity__details_messages_scroll)
        messagesView = rootView.findViewById(R.id.details_messages_table)
        formView = rootView.findViewById(R.id.new_message_area)
        sendButton = rootView.findViewById(R.id.new_message_send)
        inputField = rootView.findViewById(R.id.new_message_text)
    }

    private fun setupNewMessageForm() {
        if (User.isReadOnly()) {
            formView.hide()
            return
        }
        formView.show()

        sendButton.setOnClickListener {
            val text = inputField.text.toString().replace("\\s".toRegex(), "")
            if (text.isNotEmpty()) {
                SendMessageRequest(inputField.text.toString(), accident.id, ::sendMessageCallback)
                inputField.setText("")
            }
        }
    }

    /**
     * Обновление сообщений в списке
     */
    private fun update() {
        messagesView.removeAllViews()

        val messages = Content.messagesForAccident(accident)
        if (messages.isEmpty()) return

        var last = 0
        val group = ArrayList<Message>()
        messages.forEach {
            if (last != it.owner && last != 0) {
                addMessageRows(group)
                group.clear()
            }
            group.add(it)
            last = it.owner
        }
        addMessageRows(group)

        updateUnreadMessages()
    }

    //todo pizdets
    private fun addMessageRows(list: List<Message>) {
        if (list.size == 1) {
            drawRow(MessageRowFactory.makeOne(activity, list.first()), list.first())
        } else {
            drawRow(MessageRowFactory.makeFirst(activity, list.first()), list.first())
            (1..list.size - 2).forEach { drawRow(MessageRowFactory.makeMiddle(activity, list[it]), list[it]) }
            drawRow(MessageRowFactory.makeLast(activity, list.last()), list.last())
        }
    }

    private fun drawRow(view: View, message: Message) {
        view.setOnLongClickListener(MessageRowLongClickListener(message))
        messagesView.addView(view)
    }

    private fun updateUnreadMessages() {
        if (Content.messagesForAccident(accident).isEmpty()) return
        StoreMessages.setLast(accident.id, accident.messagesCount)
    }

    private fun sendMessageCallback(response: ApiResponse) {
        try {
            if (response.hasError()) {
                val text = response.error.text
                activity.runOnUiThread { activity.showToast(text) }
            } else {
                accident.messagesCount++
            }
        } catch (e: JSONException) {
            activity.showToast("Неизвестная ошибка" + response.toString())
            e.printStackTrace()
        }

        Content.requestDetailsForAccident(accident, {
            activity.runOnUiThread {
                (activity as AccidentDetailsActivity).update()
                update()
                scrollView.post { scrollView.fullScroll(ScrollView.FOCUS_DOWN) }
            }
        })
    }

    private inner class MessageRowLongClickListener internal constructor(private val message: Message) : View.OnLongClickListener {

        override fun onLongClick(view: View): Boolean {
            val popupWindow: PopupWindow = MessageContextMenu(activity, message)
            val viewLocation = IntArray(2)
            view.getLocationOnScreen(viewLocation)
            popupWindow.showAtLocation(view, Gravity.NO_GRAVITY, viewLocation[0], viewLocation[1])
            return true
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(ACCIDENT_ID_KEY, accident.id)
        super.onSaveInstanceState(outState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState == null) return
        accident = Content[savedInstanceState.getInt(ACCIDENT_ID_KEY)]!!
    }
}
